### ISSUE TITLE
ClientBootstrap and ServerBootstrap should accept socket options

### ISSUE THREAD

**Issue by enis** [facebook/wangle#85]
One can configure socket level options (TCP_NODELAY, etc) in Netty's Bootstrap like this: 
```
    this.channel = new Bootstrap().group(rpcClient.group).channel(rpcClient.channelClass)
        .option(ChannelOption.TCP_NODELAY, rpcClient.isTcpNoDelay())
        .option(ChannelOption.SO_KEEPALIVE, rpcClient.tcpKeepAlive)
        .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, rpcClient.connectTO)
```
At least in ServerBootstrap there is a configurable ServerSockerFactory which can be used to configure the socket. In ClientBootstrap, I could not find an easy way to do that. 

We need this for the HBase-native-client, happening here: https://issues.apache.org/jira/browse/HBASE-14850

**Comment by ChingweiYu** [comment]
Dear all,

Hi. I am a new of here but I want to have some contribution.  How can I
catch up?    I have learned C++ primer.
What I next step?

Thanks.

2017-02-14 7:54 GMT+08:00 Enis Soztutar <notifications@github.com>:

**Comment by anirudhvr** [comment]
@enis Yeah looks like there's no easy way yet. Probably cleanest would be to abstract out the common SocketConfig methods and create a separate ClientSocketConfig and ServerSocketConfig, and to modify ClientBootstrap to use a socket factory. cc @djwatson

**Comment by djwatson** [comment]
Yea pretty much.  Now accepting diffs :)
